### PR TITLE
Clean up ADAMContext

### PR DESCRIPTION
**Pull request by fnothaft**
Based on #1117.

Cleaning up `org.bdgenomics.adam.rdd.ADAMContext`:
    
* Made constructor private.
* Removed dead implicits.
* Made `loadParquet` package private.
* Broke long lines.
* Added scaladoc.
* Removed non-idiomatic uses of `Option`s.
* Removed unused overloaded `loadFeatures` definition.
* Removed `loadGenes` and `PrintGenes` command.
* Added support for .vcf.gz in `loadGenotypes` and `loadVariants`.
* Removed Java implicits conversions in ADAMContext.

I've separated that last one out into a second commit (040396e) for now. I'm OK with both squashing it or not.

### CODE REVIEW

**Comment by jpdna** [comment]
+1 to these, will plan to merge later in the day 8/22 unless more comments

**Comment by heuermh** [comment]
Would like a chance to test the glob and directory path changes.

I wonder if @akmorrow13 and @tdanford are ok with removing the gene-related functionality?

**Comment by fnothaft** [comment]
@heuermh Agreed; I'm dropping a -1 here to block the merge, since this depends on #1117 which still needs more review.

**Comment by fnothaft** [comment]
> I wonder if @akmorrow13 and @tdanford are ok with removing the gene-related functionality?

I hadn't realized until now that we're using the Gene models downstream in Mango. IMO, I think we should drop the hierarchical gene models in favor of extracting flat transcripts/exons from feature files. @heuermh I think you had suggested this a while ago. The hierarchical gene models have always had iffy performance, since they rely on something like 4 consecutive `groupBy`s. I think the flat feature approach is sufficient for what Mango is doing, but indeed, @akmorrow13 is the correct person to comment!

**Comment by akmorrow13** [comment]
By flat feature approach do you mean having an object of (transcript, genid..) without the exons? Whould the exons be a separate object?

Best,

Alyssa Morrow
akmorrow@berkeley.edu <mailto:akmorrow@berkeley.edu>
414-254-6645

**Comment by heuermh** [comment]
> By flat feature approach do you mean having an object of (transcript, genid..) without the exons? Whould the exons be a separate object?

You would have an `RDD[Feature]`, so in that sense it is flat.  The structure of the features themselves is rich, including nesting via the `parentIds` &rarr; `featureId` relationship and typing via `featureType`.

**Comment by fnothaft** [comment]
What @heuermh said. ;)
For an example, see how the `transcriptome: RDD[Feature]` is used [here](https://github.com/bigdatagenomics/quinine/blob/master/quinine-core/src/main/scala/org/bdgenomics/quinine/metrics/rna/RNASeqPanelStats.scala#L69).

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 1]
```diff
@@ -849,6 +1119,14 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
     DatabaseVariantAnnotationRDD(rdd, sd)
   }
 
+  /**
+   * Loads NucleotideContigFragments stored in Parquet, with metadata.
+   *
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: comment doesn't match method

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 2]
```diff
@@ -835,8 +1097,16 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
     FragmentRDD(rdd, sd, rgd)
   }
 
-  def loadVcfAnnotations(
-    filePath: String): DatabaseVariantAnnotationRDD = {
+  /**
+   * Loads variant annotations from a VCF.
+   *
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: Loads variant annotations stored in VCF format.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 3]
```diff
@@ -925,10 +1241,24 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
...
+   * @see loadVcf
+   * @see loadParquetGenotypes
+   */
   def loadGenotypes(
     filePath: String,
     projection: Option[Schema] = None): GenotypeRDD = {
-    if (filePath.endsWith(".vcf")) {
+    if (filePath.endsWith(".vcf") ||
+      filePath.endsWith(".vcf.gz")) {
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: Should also support `bgzf` and `bgz` file extensions

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 4]
```diff
@@ -937,10 +1267,24 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
...
+   *
+   * @see loadVcf
+   * @see loadParquetVariants
+   */
   def loadVariants(
     filePath: String,
     projection: Option[Schema] = None): VariantRDD = {
-    if (filePath.endsWith(".vcf")) {
+    if (filePath.endsWith(".vcf") ||
+      filePath.endsWith(".vcf.gz")) {
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: same as above

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala (hunk 2 above)
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: OOC, why do you see that text as preferable?

Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: Little minds, consistency, etc.  😉  Reads the same as the other similar methods.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 5]
```diff
@@ -906,6 +1207,21 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
     }
   }
 
+  /**
+   * Auto-detects the file type and loads contigs as an RDD.
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: an RDD &rarr; a NucleotideContigFragmentRDD

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala [hunk 6]
```diff
@@ -1011,6 +1355,19 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
     }
   }
 
+  /**
+   * Loads alignments from a given path, and infers the input type.
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: should this be alignment fragments?  Or better yet:  Auto-detects the file type and loads a FragmentRDD.

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala (hunk 1 above)
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: Sorry, what here doesn't match the method?

Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala]: Something is wonky with the github web UI, look at line 1136 in view whole file, it should be "Load VariantAnnotations ...".  Then line 1126 def loadParquetVariantAnnotations is missing doc.

https://github.com/fnothaft/adam/blob/27b769c69e7a2a864ba7a99d824213df72909acc/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala#L1136

**Comment by heuermh** [comment]
For #1129, do you want to leave those changes out, have me PR against this, or add on to this PR?

In addition to what you have here, we would want to remove
```
adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/GeneRDD.scala
adam-core/src/main/scala/org/bdgenomics/adam/models/Gene.scala
```

`Features.gatherAttributes` method
https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/Features.scala#L176

`FeatureRDD.rddToGeneRDD` and `FeatureRDD.toGenes` methods
https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureRDD.scala#L144

and relevant unit tests.

**Comment by fnothaft** [comment]
How soon do you think you'll have #1129 done? If by the end of the week, let's tack it on here. If it'd be next week, let's just open it whenever you're ready.

**Comment by heuermh** [comment]
It just takes application of the delete key  :)

File: adam-core/src/test/scala/org/bdgenomics/adam/rdd/ADAMContextSuite.scala [hunk 7]
```diff
@@ -422,17 +330,69 @@ class ADAMContextSuite extends ADAMFunSuite {
...
+    val path = new File(resourcePath("vcf_dir/1.vcf")).getParent()
+
+    val variants = sc.loadVcf(path).toVariantRDD
+    assert(variants.rdd.count === 681)
+  }
+
+  sparkTest("load parquet with globs") {
+    val inputPath = resourcePath("small.sam")
+    val reads = sc.loadAlignments(inputPath)
+    val outputPath = tempLocation()
```
Comment (heuermh) [inline adam-core/src/test/scala/org/bdgenomics/adam/rdd/ADAMContextSuite.scala]: This fails to compile for me on your branch.  Are you depending on a different version of utils?

**Comment by heuermh** [comment]
I've tested this and haven't run into any problems.

LGTM once the one remaining doc fix above (lines 1126 and 1136 in `AdamContext.scala`) is in.

**Comment by fnothaft** [comment]
Thanks @heuermh! Let me get those fixes and rebase this now.

**Comment by fnothaft** [comment]
Just rebased and addressed comments. Thanks @heuermh!

**Comment by heuermh** [comment]
Ready to merge?  Commits the way you'd like them?

**Comment by fnothaft** [comment]
All is good from my side! Feel free to merge, @heuermh.

**Comment by heuermh** [comment]
Great, will do!

I feel like this merge should come with a thunderbolt or some other cool sound effects.

**Comment by fnothaft** [comment]
> I feel like this merge should come with a thunderbolt or some other cool sound effects.

Could be a nice addition to the `./scripts/commit-pr.sh` script... ;)

**Pull request merged**